docker/Dockerfile

```diff
@@ -1,12 +1,23 @@
-FROM golang:alpine AS builder
-
-RUN mkdir /src
-ADD . /src
-
-RUN apk add --no-cache make git curl && \
-	cd /src && \
-	mkdir -p output/bin && \
-	make V=1 bin/step
+FROM --platform=$BUILDPLATFORM golang:alpine AS builder
+
+WORKDIR /src
+COPY go.mod go.sum .
+RUN go mod download
+COPY . .
+RUN apk add --no-cache make git curl
+
+ARG TARGETOS
+ARG TARGETARCH
+ARG TARGETVARIANT
+RUN --mount=type=cache,target=/root/.cache/go-build \
+    --mount=type=cache,target=/go/pkg \
```

Comment:
These cache mounts really only improve performance on iterative builds of the same image. They won't speed up the release, and we could remove them to keep this file simpler.

Reply:
/go/pkg has a directory `mod` with the source code of the packages, this should common for everybody, and a $GOOS_$GOARCH directory with compiled files, that are only for that specific combination, e.g. `linux_amd64`

Reply:
I experimented with something similar but wasn't building everything in the same image. I used a similar cache and a different builder image because I needed dependencies for each architecture. I didn't merge it because, at least for that project, docker was taking more time to COPY from the cache than Go downloading the packages. This was my PR https://github.com/smallstep/step-kms-plugin/pull/34